simple_graph/weighted_graph.py

```diff
@@ -0,0 +1,170 @@
+
+    def has_node(self, n):
+        """True if node 'n' is contained in the graph, False if not."""
+        return n in self.dict
+
+    def add_edge(self, n1, n2, weight=0):
+        """Adds a new edge to the graph connecting 'n1' to 'n2', if either n1
+        or n2 are not already present in the graph, they should be added."""
+        self.add_node(n1)
+        self.add_node(n2)
+        self.dict[n1][n2] = weight
+
+    def del_node(self, n):
+        """Delete 'n' from the graph, raise error if node doesn't exist."""
+        try:
+            del self.dict[n]
+            # remove edges pointing to n
+            for key, value in self.dict.iteritems():
+                if n in value:
+                    del self.dict[key][n]
+        except (ValueError, KeyError):
+                raise AttributeError('No Such Node Exists')
+
+    def del_edge(self, n1, n2):
+        """Delete edge connecting 'n1' and 'n2' from the graph, raise
+        error if edge doesn't exist."""
+        try:
+            del self.dict[n1][n2]
+        except (ValueError, KeyError):
+            raise AttributeError('No Such Edge Exists')
+
+    def neighbors(self, n):
+        """Returns the list of all nodes with edges leading to them from n,
+        raise error if n is not in g."""
+
+        try:
+            return self.dict[n].keys()
+        except KeyError:
+            raise KeyError('Node not in graph.')
+
+    def adjacent(self, n1, n2):
+        """returns True if there is an edge connecting n1 and n2, False if not,
+        raises an error if either of the supplied nodes are not in g."""
+        try:
+            return n2 in self.dict[n1]
+        except KeyError:
+            raise KeyError('Node(s) not in graph.')
+
+    @timed_func
+    def depth_first_traversal(self, start):
+        """Perform a full depth-first traversal of the graph beginning at start.
+        Return the path when complete."""
+        return self.recursive_dft(start, [])
+
+
+    def recursive_dft(self, start, visited=[]):
+        """Recursive function for depth first traversal."""
+        if start not in visited:
+            visited.append(start)
+            for i in self.neighbors(start):
+                self.recursive_dft(i, visited)
+            return visited
+
+    @timed_func
+    def breadth_first_traversal(self, start):
+        """Perform a full breadth-first traversal of the graph beginning at start.
+        Return the path when complete."""
+        visited = []
+        visited.append(start)
+        start_visited = visited
+        while True:
+            temp = []
+            for node_ in start_visited:
+                for i in self.neighbors(node_):
+                    if i not in visited:
+                        visited.append(i)
+                        temp.append(i)
+            start_visited = temp
+            if not temp:
+                break
+        return visited
+
+
+def populated_graph():
```

Comment:
Do you use this function outside of the `if __name__ == "__main__":` block?  If not, why define it at module level?